### PR TITLE
Added specific namespace and service account to PV recycler

### PR DESCRIPTION
**Pull request by markturansky**
Resolves https://github.com/openshift/origin/issues/6883

Added namespace and SA to recycler to fix permission issues.  Also updated the Upgrade doc.

@pweil- @deads2k

### CODE REVIEW

File: UPGRADE.md [hunk 1]
```diff
@@ -81,3 +81,4 @@ references:
 ## Origin 1.1.x / OSE 3.1.x
 
 1. The `buildconfig` label on Build objects, which contains the name of the BuildConfig for the Build, has been deprecated in favor of a new `openshift.io/build-config.name` label.
+2. New Service Accounts specific to the Persistent Volume operations of binding, recycling, and provisioning were added.  Run `oadm policy reconcile-sccs --confirm` to update your Security Contexts and generate the new Service Accounts.
```
Comment (liggitt) [inline UPGRADE.md]: 'service accounts'
'PersistentVolume'
'SecurityContextConstraints'
this doesn't generate the new service accounts

File: pkg/cmd/server/bootstrappolicy/securitycontextconstraints.go [hunk 2]
```diff
@@ -241,7 +241,7 @@ func GetBoostrapSCCAccess(infraNamespace string) (map[string][]string, map[strin
 	}
 
 	buildControllerUsername := serviceaccount.MakeUsername(infraNamespace, InfraBuildControllerServiceAccountName)
-	pvControllerUsername := serviceaccount.MakeUsername(infraNamespace, InfraPersistentVolumeBinderControllerServiceAccountName)
```
Comment (liggitt) [inline pkg/cmd/server/bootstrappolicy/securitycontextconstraints.go]: s/pvControllerUsername/pvRecyclerControllerUsername/

File: pkg/cmd/server/kubernetes/master.go [hunk 3]
```diff
@@ -103,9 +103,11 @@ func (c *MasterConfig) RunPersistentVolumeProvisioner(client *client.Client) {
 	}
 }
 
-func (c *MasterConfig) RunPersistentVolumeClaimRecycler(recyclerImageName string, client *client.Client) {
+func (c *MasterConfig) RunPersistentVolumeClaimRecycler(recyclerImageName string, client *client.Client, serviceAccountName string, namespace string) {
 	uid := int64(0)
 	defaultScrubPod := volume.NewPersistentVolumeRecyclerPodTemplate()
+	defaultScrubPod.Namespace = namespace
+	defaultScrubPod.Spec.ServiceAccountName = serviceAccountName
```
Comment (liggitt) [inline pkg/cmd/server/kubernetes/master.go]: The controller is running as the service account, so it has permission whether or not the pod's service account does. I don't really want the recycler pod having an API token that lets it do all the things the controller can do

File: pkg/cmd/server/start/start_master.go [hunk 4]
```diff
@@ -558,7 +558,7 @@ func startControllers(oc *origin.MasterConfig, kc *kubernetes.MasterConfig) erro
 		kc.RunNamespaceController()
 		kc.RunPersistentVolumeClaimBinder(binderClient)
 		kc.RunPersistentVolumeProvisioner(provisionerClient)
-		kc.RunPersistentVolumeClaimRecycler(oc.ImageFor("recycler"), recyclerClient)
+		kc.RunPersistentVolumeClaimRecycler(oc.ImageFor("recycler"), recyclerClient, bootstrappolicy.InfraPersistentVolumeRecyclerControllerServiceAccountName, oc.Options.PolicyConfig.OpenShiftInfrastructureNamespace)
```
Comment (liggitt) [inline pkg/cmd/server/start/start_master.go]: above, we build the recyclerClient to use the recycler controller service account already... you don't need to also give the pod the powers of that service account

Comment (markturansky) [inline pkg/cmd/server/start/start_master.go]: I added that before fixing the namespace.  I'll test/verify it's not needed and then remove it.

File: UPGRADE.md (hunk 1 above)
Comment (markturansky) [inline UPGRADE.md]: comments changed

File: pkg/cmd/server/bootstrappolicy/securitycontextconstraints.go (hunk 2 above)
Comment (markturansky) [inline pkg/cmd/server/bootstrappolicy/securitycontextconstraints.go]: var name changed

File: pkg/cmd/server/start/start_master.go (hunk 4 above)
Comment (markturansky) [inline pkg/cmd/server/start/start_master.go]: testing, verified, and removed.

File: pkg/cmd/server/start/start_master.go [hunk 5]
```diff
@@ -558,7 +558,7 @@ func startControllers(oc *origin.MasterConfig, kc *kubernetes.MasterConfig) erro
 		kc.RunNamespaceController()
 		kc.RunPersistentVolumeClaimBinder(binderClient)
 		kc.RunPersistentVolumeProvisioner(provisionerClient)
-		kc.RunPersistentVolumeClaimRecycler(oc.ImageFor("recycler"), recyclerClient)
+		kc.RunPersistentVolumeClaimRecycler(oc.ImageFor("recycler"), recyclerClient, oc.Options.PolicyConfig.OpenShiftInfrastructureNamespace)
```
Comment (liggitt) [inline pkg/cmd/server/start/start_master.go]: do we need to make sure the OpenShiftInfrastructureNamespace isn't subject to the cluster-wide default node selector?

**Comment by markturansky** [comment]
Squashed

**Comment by markturansky** [comment]
Thanks for the review @liggitt

**Comment by markturansky** [comment]
@liggitt any reason this can't be merged?

**Comment by liggitt** [comment]
Hadn't thought through the implications of moving the recycler pod from the default namespace to the openshift-infra namespace

**Comment by smarterclayton** [comment]
This is causing e2e persistent volume tests to fail, which is good (because they would have caught it) and bad (because we weren't running them).  Will turn them on after this merges.

**Comment by liggitt** [comment]
LGTM, though I'd like @detiber or @sdodson to weigh in on running the recycler pod in the openshift-infra namespace instead of the default namespace (in case the installer sets up project nodeselectors any differently for those two projects)

**Comment by sdodson** [comment]
> LGTM, though I'd like @detiber or @sdodson to weigh in on running the recycler pod in the openshift-infra namespace instead of the default namespace (in case the installer sets up project nodeselectors any differently for those two projects)

We don't currently setup project node selectors out of the box.

**Comment by liggitt** [comment]
Rebase (and is there an extended test that can be run for the recycler)

**Comment by smarterclayton** [comment]
There is an extended test, NFS, which is failing due to this

**Comment by markturansky** [comment]
Rebased

**Comment by liggitt** [comment]
LGTM, [merge]

**Comment by smarterclayton** [comment]
- FAIL: TestBootstrappedConstraints (0.00s)
	securitycontextconstraints_test.go:35: unexpected user access for
hostmount-anyuid.  Found
[system:serviceaccount:openshift-infra:pv-recycler-controller], wanted
[system:serviceaccount:openshift-infra:pv-binder-controller]
--- FAIL: TestBootstrappedConstraintsWithAddedUser (0.00s)
	securitycontextconstraints_test.go:59: unexpected user access for
hostmount-anyuid.  Found
[system:serviceaccount:openshift-infra:pv-recycler-controller], wanted
[system:serviceaccount:openshift-infra:pv-binder-controller]
FAIL



Didn't even pass the unit tests?

**Comment by markturansky** [comment]
Looking into it. This PR has been all green at some point.

**Comment by markturansky** [comment]
Test fixed. Rebased for good measure.

**Comment by liggitt** [comment]
[merge]

**Comment by liggitt** [comment]
actually, squash before merge

**Comment by markturansky** [comment]
squashed

**Comment by liggitt** [comment]
[merge]

**Comment by markturansky** [comment]
I rebased for good measure.  tests run locally for me.  I'm not sure what the problem with the bot is.

**Comment by liggitt** [comment]
[merge]

**Comment by markturansky** [comment]
all green this time

**Pull request merged**